DR_0008_role-based-permissions.md

```diff
@@ -0,0 +1,187 @@
+
+##### Organization View Mixins
+The `ProjectQuerySetMixin` has been removed and replaced with `ProjectListMixin` to provide role-based filtering of projects for `UI` and `API` project list views.  Other than this there are minimal changes to organization mixins.
+
+##### Views
+
+The existing `django-tutelary` approach to determining the required permissions for a view has been replaced with a more conventional django approach. Views can set permissions in one of three ways:
+1. Set the permission using the `permission_required` class attribute.
+2. Set `permission_required` to a dict which maps a single permission to a request method.
+3. Override `get_permission_required` if a more complex approach to determining permissions is needed, e.g. if different permissions are required based on the request method and/or organization or project visibility.
+
+##### Code examples
+
+The following code examples are indicative of the approach and subject to change:
+
+##### Groups and Permissions
+
+We define a set of platform-wide Permissions, eg.,
+
+```python
+p1 = Permission(
+		name=’View Project’, content_type=’Project’, codename=’project.view’
+	 )
+```
+
+We create the following platform-wide Groups:
+
+`PublicUser`, `ProjectUser`, `DataCollector`, `ProjectManager`, `OrgAdmin`, `SuperUser` eg:
+
+```python
+project_user = Group(name=’ProjectUser’, permissions=[p1,...,pn])
+```
+
+##### Roles
+
+We define an abstract base model Role with the following attributes:
+
+```python
+class Role (model.Model):
+	class Meta:
+	    abstract = True
+		user = ForeignKey(‘accounts.User’)
+
+	@cached_property
+	def permissions(self):
+		return [perm.codename
+				for perm in self.group.permissions.all()]
+```
+
+We then modify `OrganizationRole` and `ProjectRole` to extend `Role`:
+
+```python
+class OrganizationRole(Role):
+	organization = ForeignKey(Organization)
+	group = models.ForeignKey(
+        'auth.Group', related_name='organization_roles')
+
+class ProjectRole(Role):
+	project = ForeignKey(Project)
+	role = CharField(choices=ROLE_CHOICES, default=’PU’)
+	group = models.ForeignKey(
+        'auth.Group', related_name='project_roles')
+```
+
+
+When a Role is created it is associated with its corresponding Group, eg:
+
+```python
+pu = Group.objects.get(name='ProjectUser')
+role = ProjectRole(user=user, group=pu,
+				   project=project, role=’PU’)
```

Comment:
This definitely falls under the "nitpick" category, but there are a number of whitespace issues in the above examples that slightly hamper the readability.